edk2toollib/uefi/edk2/build_objects/dsc.py

```diff
@@ -0,0 +1,455 @@
+        if type(other) is not dsc:
+            return False
+        if other.skus != self.skus:
+            return False
+        if other.components != self.components:
+            return False
+        if other.library_classes != self.library_classes:
+            return False
+        if other.build_options != self.build_options:
+            return False
+        if other.pcds != self.pcds:
+            return False
+        if other.defines != self.defines:
+            return False
+        return True
+
+
+class dsc_section_type:
+    dsc_module_types = ["COMMON", "BASE", "SEC", "PEI_CORE", "PEIM", "DXE_CORE", "DXE_DRIVER",
+                        "DXE_RUNTIME_DRIVER", "DXE_SAL_DRIVER", "DXE_SMM_DRIVER", "SMM_CORE", "UEFI_DRIVER", "UEFI_APPLICATION", "USER_DEFINED"]
+
+    def __init__(self, arch="common", module_type="common"):
+        self.arch = arch.upper().strip()
+        self.module_type = module_type.upper().strip()
+        if not dsc_section_type.IsValidModuleType(self.module_type):
+            raise ValueError(f"{module_type} is not a proper module type for dsc section")
+
+    def __hash__(self):
+        arch = "*" if (self.arch == "COMMON" or self.arch == "DEFAULT") else self.arch
+        return hash((arch, self.module_type))
+
+    def __eq__(self, other):
+        if type(other) is not dsc_section_type:
+            return False
+        arch = "*" if (self.arch == "COMMON" or self.arch == "DEFAULT") else self.arch
+        arch2 = "*" if (other.arch == "COMMON" or other.arch == "DEFAULT") else other.arch
+        return self.module_type == other.module_type or arch == arch2
+
+    def __repr__(self):
+        attributes = f".{self.arch}.{self.module_type}"
+        return attributes
+
+    @classmethod
+    def IsValidModuleType(cls, name):
+        return name in cls.dsc_module_types
+
+
+class dsc_buildoption_section_type(dsc_section_type):
+    '''
+    [BuildOptions.$(arch).CodeBase.Edk2ModuleType]
+    [BuildOptions.$(arch).CodeBase]
+    [BuildOptions.common.CodeBase]
+    [BuildOptions.$(arch)]
+    [BuildOptions.common]
+    [BuildOptions]
+    '''
+
+    def __init__(self, arch="common", codebase="common", module_type="common"):
+        super().__init__(arch, module_type)
+        self.codebase = codebase.upper().strip()
+        if not self.IsValidCodeBase(self.codebase):
+            raise ValueError(f"{codebase} is not a valid codebase type")
+
+    @classmethod
+    def IsValidCodeBase(cls, codebase):
+        return codebase in ["COMMON", "EDK", "EDKII"]
+
+    def __hash__(self):
+        return hash((super().__hash__(), self.codebase))
+
+    def __eq__(self, other):
+        if type(other) is not dsc_buildoption_section_type:
+            return False
+        return other.module_type == self.module_type and other.codebase == self.codebase and other.module_type == self.module_type
+
+    def __repr__(self):
+        return f"{self.arch}.{self.codebase}.{self.module_type}"
+
```

Comment:
Like dsc_pcd_componet_type, should there be a dsc_buildoption_component_type?

Reply:
We don't need it because PCD's is the only thing that's a dictionary rather than a set in the component class. You can't set and sort of identifier for the buildoption type